tests/integration2/pilot/sidecar_api_test.go

```diff
@@ -0,0 +1,134 @@
+package pilot
+
+import (
+	"fmt"
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/api/components"
+	"istio.io/istio/pkg/test/framework/api/descriptors"
+	"istio.io/istio/pkg/test/framework/api/ids"
+	"istio.io/istio/pkg/test/framework/api/lifecycle"
+	"istio.io/istio/pkg/test/framework/runtime/components/environment/native"
+	"istio.io/istio/pkg/test/scopes"
+	"istio.io/istio/pkg/test/util/structpath"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+func TestSidecarListeners(t *testing.T) {
+	// Call Requires to explicitly initialize dependencies that the test needs.
+	ctx := framework.GetContext(t)
+	scopes.Framework.SetOutputLevel(log.DebugLevel)
+
+	// TODO: Limit to Native environment until the Kubernetes environment is supported in the Galley
+	// component
+	ctx.RequireOrSkip(t, lifecycle.Test, &descriptors.NativeEnvironment)
+	ctx.RequireOrFail(t, lifecycle.Test, &ids.Galley)
+	ctx.RequireOrFail(t, lifecycle.Test, &ids.Pilot)
+
+	// Load the canonical dataset into Galley and by implication Pilot
+	gal := components.GetGalley(ctx, t)
+	env, err := native.GetEnvironment(ctx)
+	components.GetMixer(ctx, t)
+	s := strings.Split(env.Mesh.MixerCheckServer, ":")[1]
+	mixerPort, _ := strconv.ParseUint(s, 10, 32)
+
+	path, err := filepath.Abs("../../testdata/config")
+	if err != nil {
+		t.Fatalf("No such directory: %v", err)
+	}
+	gal.ApplyConfigDir(path)
+
+	pilot := components.GetPilot(ctx, t)
+
+	// Simulate proxy identity of egress gateway ...
+	nodeId := &model.Proxy{
+		ClusterID:       "integration-test",
+		Type:            model.SidecarProxy,
+		IPAddresses:     []string {"10.2.0.1"},
+		ID:              "app3.testns",
+		DNSDomain:       "testns.cluster.local",
+	}
+
+	// ... and get listeners from Pilot for that proxy
+	req := &xdsapi.DiscoveryRequest{
+		Node: &core.Node{
+			Id:  nodeId.ServiceNode(),
+		},
+		TypeUrl: "type.googleapis.com/envoy.api.v2.Listener",
+	}
+	resp, err := pilot.CallDiscovery(req)
```

Comment:
I think you want this to be in a loop. Otherwise, you're going to run into timing issues.
https://github.com/istio/istio/blob/cf7fc4b7170b8c7d1b87e5ee3b5c9a1bc8001ea8/tests/integration2/galley/conversion/conversion_test.go#L98

https://github.com/istio/istio/blob/cf7fc4b7170b8c7d1b87e5ee3b5c9a1bc8001ea8/pkg/test/framework/runtime/components/galley/client.go#L64

Reply:
@nmittler is waiting on galley sufficient, after all this test is checking Pilot behavior. Does this cover the ACK from Pilot over MCP ?

@rshriram @costinm thoughts about how to test for quiesence ?

Reply:
FWIW, I didn't mean to wait on that particular Galley-based method. It is an example for quiescence. I think we should be able to use a similar method for quiescence.

Reply:
/cc @smawson 

Sure, but the issue is what to poll here so we can detect that a change transaction has propagated through the system so that a test can proceed.  From a test writer perspective I'd like to be able to do something like ...

marker := galleyComponent.ApplyConfig(<some config>)
pilotComponent.awaitMarker(marker, <timeout>)
envoyComponent.awaitMarker(marker, <timeout>)

Such a system requires that the component driver can create ids that propagate through MCP & xDS to downstream components that can then be queried. Anything else like relying on varz that track resource update counts on the components would be brittle.

Reply:
Spitballing here a bit. We can use unique/nonce (K8s) annotations, which should carry over MCP. However, I don't think there is a prevalent counterpart in xDS (There is core.metadata field, but it doesn't look like it is embedded in all of the resources). We can also try mangling resources names with nonce markers, but it makes debugging things harder. (In either case, Pilot needs to transform and ferry these over to the xDS side somehow).
Even if we had that, it leaves a couple of problems on table. There is really no transaction boundary in the system that you can reliably use to say "all of the configuration in this batch is received by the other side". This is a problem for both API Server=>Galley=>MCP and  MCP=>Pilot=>xDS.

Reply:
Thinking a bit more. It is a bit heavy-handed but: We can possibly use Ctrlz for this purpose, in Pilot and Galley (and configdump in Envoy?). If each component in-line can expose the FQN(&Version) of all input resources it has processed, and output FQNs it is serving for that input, it creates a good enough provenance chain for a test framework to track.

Reply:
AFAICT, Pilot currently doesn't provide provenance from input (MCP) to output (xDS). VersionInfo would be a more appropriate place IMO to propagate this information. Nonce is more for pairing request/responses. 

Pilot currently uses the current time concatenated with monotonically increasing integer [here](https://github.com/istio/istio/blob/0d51688e92b5176d77a4d396b7c49e8e7974cae9/pilot/pkg/proxy/envoy/v2/discovery.go#L328-L331) in the VersionInfo. For the purposes of testing, the VersionInfo could be extended to include a continuation of all MCP source VersionInfo. Galley would need to provide a mapping of VersionInfo (per collection) to the individual resources in that. The latter could be exposed via CRD status or via some out-of-band diagnostic endpoint (e.g. ctrlz). I think that would address the k8s->galley->pilot path. A similar exercises would be needed to include envoy in this path.

Alternatively, we've discussed using a purely time based scheme for VersionInfo. e.g. Pilot output after time T1 is assumed to include resources from Galley if Pilot ACK'd those resources (via MCP) before T1. This requires some assumptions on time skew and tolerance across compute boundaries.

Reply:
thats probably what Ill do but this will vary wildly across tests and folks will get it wrong. standardizing it would make folks more effective

Reply:
Updated the test to do this. Form is now...

- start stream
- wait for ADS even that matches acceptance criteria, then test it
- ...